Bluetooth: controller: Add LE Read Buffer Size v2 support

cvinayak opened the pull request:
Add LE Read Buffer Size v2 support in the controller
implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Add LE Read Buffer Size v2 support

Add LE Read Buffer Size v2 support in the controller
implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

jhedberg reviewed (changes requested):
Looks fine, but could you split the `hci.h` renames which cause updates to the host side into a separate commit? (it's not really proper to have host changes in a commit starting with "Bluetooth: controller: ..")

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1156,6 +1162,32 @@ static void le_read_buffer_size(struct net_buf *buf, struct net_buf **evt)
+#else
+#define BT_CTLR_ISO_TX_BUFFERS 0
+#endif
+#if defined(CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE)
+#define BT_CTLR_ISO_TX_BUFFER_SIZE CONFIG_BT_CTLR_ISO_TX_BUFFER_SIZE
+#else
+#define BT_CTLR_ISO_TX_BUFFER_SIZE 0
+#endif
```
Why is this necessary? `BT_CTLR_ISO_TX_BUFFER_SIZE` is not conditional on anything, so it's always defined.

cvinayak replied:
oh, `BT_CTLR_ISO_TX_BUFFER_SIZE` should have a `depends on`, will add it.

cvinayak replied:
I have remove the `if defined(...` checks and added the dependencies in Kconfig. 

Not going to do anything for ACL buffers wrt `CONFIG_BT_CONN` there is a different PR that removes controller Kconfig related to ACL Tx count and buffer size. 

I will rebase this PR after #34771 is merged.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -1156,6 +1162,32 @@ static void le_read_buffer_size(struct net_buf *buf, struct net_buf **evt)
 	rp->le_max_num = CONFIG_BT_CTLR_TX_BUFFERS;
 }
 
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
```
@Thalley Should this be as below? Does host built for only Synchronised Receiver use LE Read Buffer Size v2 command?

```suggestion
#if defined(CONFIG_BT_CTLR_ISO)
```

cvinayak replied:
ok

joerchan commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2423,27 +2423,27 @@ static void read_buffer_size_v2_complete(struct net_buf *buf)
 		bt_dev.le.acl_mtu);
 
-	max_num = MIN(rp->acl_max_pkt, CONFIG_BT_CONN_TX_MAX);
+	max_num = MIN(rp->acl_max_num, CONFIG_BT_CONN_TX_MAX);
```
Unrelated to this change.

@Thalley Why is it min with CONN_TX_MAX here? It is not like that for v1, so looks quite suspicious.

Thalley replied:
@joerchan I guess in case of a non-Zephyr controller, we might expect `CONFIG_BT_CONN_TX_MAX` in the host, but but will need to handle it if the controller returns a smaller value, or even vice versa. 

If controller responds with 5 and CONFIG_BT_CONN_TX_MAX=10, then the maximum number of buffers we can use is 5, and vice versa if the controller responds with 10 and CONFIG_BT_CONN_TX_MAX=5, then we have only allocated 5 buffers in the host I suppose.

cvinayak committed:
Bluetooth: HCI: Rename LE Read Buffer Size v2 struct members

Rename LE Read Buffer Size v2 struct members to related to
Bluetooth Specification use Length and Number terminology.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3248,6 +3280,12 @@ static int controller_cmd_handle(uint16_t  ocf, struct net_buf *cmd,
 		le_read_buffer_size(cmd, evt);
 		break;
 
+#if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
+	case BT_OCF(BT_HCI_OP_LE_READ_BUFFER_SIZE_V2):
+		le_read_buffer_size_v2(cmd, evt);
+		break;
+#endif /* CONFIG_BT_CTLR_ADV_ISO || CONFIG_BT_CTLR_CONN_ISO */
```
LE Read Buffer Size v2 is needed only when ISO feature is supported.

Cannot use `IS_ENABLED()` in switch-case, else will need to use `else` that returns unsupported command return code. Buy may be compiler is intelligent to optimize the else path?

Thalley replied:
Ah, right, that's a good point. Didn't consider that :)

cvinayak committed:
Bluetooth: controller: Add ISO Buffer size Kconfig dependency

Add BT_CTLR_ISO_TX_BUFFER_SIZE depends on Broadcast ISO
and/or Connected ISO feature being enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Minor conditional compile comment correction

Corrected a conditional compile comment and added some
newlines.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented:
@jhedberg can you please take another look?
